presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java

```diff
@@ -37,11 +41,48 @@ public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArgu
 
     public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArguments, MethodHandle methodHandle, Optional<MethodHandle> instanceFactory, boolean deterministic)
     {
+        this(nullable, nullableArguments, methodHandle, instanceFactory, deterministic, false, Optional.empty());
+    }
+
+    public ScalarFunctionImplementation(
+            boolean nullable,
+            List<Boolean> nullableArguments,
+            MethodHandle methodHandle,
+            Optional<MethodHandle> instanceFactory,
+            boolean deterministic,
+            boolean returnValueAsParameter,
+            Optional<Integer> returnValueSliceLength)
+    {
         this.nullable = nullable;
         this.nullableArguments = ImmutableList.copyOf(requireNonNull(nullableArguments, "nullableArguments is null"));
         this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
         this.instanceFactory = requireNonNull(instanceFactory, "instanceFactory is null");
         this.deterministic = deterministic;
+        this.returnValueAsParameter = returnValueAsParameter;
+        this.returnValueSliceLength = returnValueSliceLength;
+
+        int expectedParametersCount = methodHandle.type().parameterCount();
+        if (returnValueAsParameter) {
+            expectedParametersCount--;
+        }
+        if (instanceFactory.isPresent()) {
+            expectedParametersCount--;
+        }
+        if (methodHandle.type().parameterCount() >= 1 && methodHandle.type().parameterType(0) == ConnectorSession.class) {
```

Comment:
As far as I understand we don't force any other "special" parameter to be on fixed position. Maybe we can break this requirement here as well (or force this requirement on other special arguments).